### ISSUE TITLE
Tree scope for Multiple Parents per Node?

### ISSUE THREAD

**Issue by daninfocus** [staudenmeir/laravel-adjacency-list#109]
Hi, I'm having an issue where the ancestors() method creates an infinite loop (I think).
I have a pedigree database with ~100000 entries.

Im trying to get up-to 5 generations or (depth)
Dogs::find($id)->ancestors()->whereDepth('>', -5)->get(); 

I understand the whereDepth() function only filters, and does not stop the query.
And I see that in  One Parent per Node there is tree scopes that allow you to provide a maximum depth.
My question is, is there any way to specify the maximun depth for Many to Many relationships?

**Comment by staudenmeir** [comment]
Please share the `Dogs` model.

**Comment by daninfocus** [comment]
`class Dog extends Model
{
    use \Staudenmeir\LaravelAdjacencyList\Eloquent\HasGraphRelationships;

    public function enableCycleDetection(): bool
    {
        return true;
    }
    
    public function getPivotTableName(): string
    {
        return 'dogs_pivot';
    }

    public function getParentKeyName(): string
    {
        return 'parent_id';
    }

    public function getChildKeyName(): string
    {
        return 'child_id';
    }
}`

` Schema::create('dogs_pivot', function (Blueprint $table) {
            $table->increments('id');
            $table->unsignedInteger('child_id');
            $table->unsignedInteger('parent_id')->nullable()->default(null);
            $table->timestamps();
        });`

`Schema::create('dogs', function (Blueprint $table) {
            $table->id();
            $table->string('name');
            $table->string('sire_id')->nullable();
            $table->string('dam_id')->nullable();
        });`
So a dog will have 2 parent_id, one for the sire_id and another for dam_id

**Comment by staudenmeir** [comment]
I'm a bit confused about your table structure: Do you use `child_id` & `parent_id` or `sire_id` & `dam_id` to define the family tree (or both)?

**Comment by daninfocus** [comment]
Sorry, so in the pivot table parent_id is used for both sire_id and dam_id from the dogs table. 
Child_id is the dog.id, this dogs has 2 parents( sire_id and dam_id ), in the pivot table every dog.id will have 2 entries for different.
I got the idea [here](https://stackoverflow.com/questions/11372248/inbreeding-immune-database-structure):

**Comment by staudenmeir** [comment]
> where the ancestors() method creates an infinite loop (I think).

How does this show? Does the query/app run out of memory?

> I understand the whereDepth() function only filters, and does not stop the query.
> My question is, is there any way to specify the maximun depth for Many to Many relationships?

Use a [recursive query constraint](https://github.com/staudenmeir/laravel-adjacency-list#graphs-recursive-query-constraints) to actually stop the query at the maximum depth:

```php
$ancestors = Dog::withRecursiveQueryConstraint(function (Builder $query) {
   $query->whereDepth('>', -5);
}, function () {
   return Dogs::find($id)->ancestors;
});
```

I'm still working on that part and will probably add a helper like `Dog::withMaxDepth()`.

**Comment by daninfocus** [comment]
Okay thank you so much! 

> How does this show? Does the query/app run out of memory?

I have the database on WSL2 Ubuntu, and when I call the ancestors() relationship without any filters, the process for WSL uses 99% of RAM and doesn't stop, so I have to shut down the database to stop this.

**Comment by staudenmeir** [comment]
Weird, that's exactly what cycle detection should prevent.

Does it work with the depth constraint?

Is it possible for you to share a database dump of these two tables?

**Comment by daninfocus** [comment]
Yes the depth constraint works fine, no "infinite loop".
I cant share the database right now, but the structure is like this:
Dogs table:
| id  | sire_id | dam_id |
| ------------- | ------------- | ------------- |
| 1  | 2  | 3  |
| 2  | 4  | 5 |
| 3  | 4  | 5 |
| 4  | null  | null  |
| 5  | null  | null  |

Dogs Pivot table:

| id  | child_id | parent_id |
| ------------- | ------------- | ------------- |
| 1  | 1  | 2  |
| 2  | 1  | 3 |
| 3  | 2  | 4 |
| 4  | 2 | 5  |
| 5  | 3 | 4  |
| 6  | 3 | 5  |

I have a seeder that will get all the dogs parent_ids and insert them in the pivot table.
Like I said earlier about ~150000 dogs so ~300000 rows in the pivot table.

And thank you again for this wonderful package.

**Comment by staudenmeir** [comment]
I released `v1.11.1` with a `withMaxDepth()` helper:

```php
$ancestors = Dog::withMaxDepth(-5, function () {
   return Dogs::find($id)->ancestors;
});
```

**Comment by daninfocus** [comment]
Wow this is perfect thank you so much!

**Comment by CoderPlusFromNowhere** [comment]
> I released `v1.11.1` with a `withMaxDepth()` helper:
> 
> ```
> $ancestors = Dog::withMaxDepth(-5, function () {
>    return Dogs::find($id)->ancestors;
> });
> ```

I can't use withMaxDepth: I got the following error:

 Call to undefined method App\Models\User::withMaxDepth()

$users = User::withMaxDepth(4, function() {
            return User::find(Auth::user()->id)->descendants;
 });
